File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.cpp
class Solution {
public:
    int firstStableIndex(vector<int>& nums, int k) {
        int maxSoFar = INT_MIN;
        int cand = 0, cMax = 0;
        int n = nums.size();

        for (int i = 0; i < n; i++) {
            maxSoFar = max(maxSoFar, nums[i]);

            if (i == cand)
                cMax = maxSoFar;

            if (nums[i] < cMax - k)
                cand = i + 1;
        }

        return cand < n ? cand : -1;
    }
};

File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        max_so_far = -10**9
        cand = 0
        c_max = 0
        n = len(nums)

        for i in range(n):
            max_so_far = max(max_so_far, nums[i])

            if i == cand:
                c_max = max_so_far

            if nums[i] < c_max - k:
                cand = i + 1

        return cand if cand < n else -1